jorldy/run_mode.py

```diff
@@ -295,38 +292,54 @@ def async_distributed_train(config_path, unknown):
             agent.load(config.train.load_path)
 
         save_path = path_queue.get()
-        step, _step, print_stamp, save_stamp = 0, 0, 0, 0
-        while step < config.train.run_step:
-            transitions = []
-            while (_step == 0 or not trans_queue.empty()) and (
-                _step - step < config.train.update_period
-            ):
-                _step, _transitions = trans_queue.get()
-                transitions += _transitions
-            delta_t = _step - step
-            print_stamp += delta_t
-            save_stamp += delta_t
-            step = _step
-            result = agent.process(transitions, step)
+        heap = {
+            "step": 0,
+            "run_step": config.train.run_step,
+            "print_stamp": 0,
+            "save_stamp": 0,
+            "wait_thread": False,
+            "wait_process": True,
+        }
+        step, transitions = 0, []
+        print_signal, save_signal = False, False
+        gath_thread = Thread(
+            target=gather_thread,
+            args=(trans_queue, transitions, heap, "+="),
+        )
+        gath_thread.start()
+        while step < heap["run_step"]:
+            while heap["wait_process"]:
+                time.sleep(0.1)
+            heap["wait_thread"] = True
+            if step != heap["step"]:
+                step = heap["step"]
+                _transitions = transitions.copy()
+                transitions.clear()
+                is_over = step >= heap["run_step"]
+                if heap["print_stamp"] >= config.train.print_period or is_over:
+                    print_signal = True
+                    heap["print_stamp"] = 0
+                if heap["save_stamp"] >= config.train.save_period or is_over:
+                    save_signal = True
+                    heap["save_stamp"] = 0
+            heap["wait_thread"] = False
+            result = agent.process(_transitions, step)
```

Comment:
As a result of testing, the number of times to call "process" is greater than the number of times to  transitions "get(copy)".
Non-updated transitions are being used in the process method.